File: app/src/main/java/com/stwpower/powertap/data/api/MyApiClient.java
package com.stwpower.powertap.data.api;

import android.util.Log;

import java.util.concurrent.TimeUnit;

import com.google.gson.Gson;
import com.stripe.stripeterminal.external.models.ConnectionTokenException;
import com.stwpower.powertap.domain.MyResponse;
import com.stwpower.powertap.utils.LoggingInterceptor;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.Map;

import okhttp3.OkHttpClient;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import javax.net.ssl.*;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;

public class MyApiClient {
    private static final OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(20, TimeUnit.SECONDS)
            .writeTimeout(20, TimeUnit.SECONDS)
            .addInterceptor(new LoggingInterceptor())
            .build();
    private static final Retrofit mRetrofit = new Retrofit.Builder()
            .baseUrl(MyApp.baseUrl + "/")
//            .baseUrl("http://192.168.5.142:8082/power_bank/")
            .client(getUnsafeOkHttpClient())
            .addConverterFactory(GsonConverterFactory.create())
            .build();
    private static final MyApiService mService = mRetrofit.create(MyApiService.class);

    public static OkHttpClient getUnsafeOkHttpClient() {
        try {
            // Create a trust manager that does not validate certificate chains
            final TrustManager[] trustAllCerts = new TrustManager[] {
                    new X509TrustManager() {
                        @Override
                        public void checkClientTrusted(java.security.cert.X509Certificate[] chain, String authType) throws CertificateException {
                        }

                        @Override
                        public void checkServerTrusted(java.security.cert.X509Certificate[] chain, String authType) throws CertificateException {
                        }

                        @Override
                        public java.security.cert.X509Certificate[] getAcceptedIssuers() {
                            return new X509Certificate[]{};
                        }
                    }
            };

            // Install the all-trusting trust manager
            final SSLContext sslContext = SSLContext.getInstance("SSL");
            sslContext.init(null, trustAllCerts, new java.security.SecureRandom());

            // Create an ssl socket factory with our all-trusting manager
            final SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();

            OkHttpClient.Builder builder = new OkHttpClient.Builder()
                    .connectTimeout(30, TimeUnit.SECONDS)
                    .readTimeout(20, TimeUnit.SECONDS)
                    .writeTimeout(20, TimeUnit.SECONDS)
                    .addInterceptor(new LoggingInterceptor());
            builder.sslSocketFactory(sslSocketFactory, (X509TrustManager)trustAllCerts[0]);
            builder.hostnameVerifier(new HostnameVerifier() {
                @Override
                public boolean verify(String hostname, SSLSession session) {
                    return true;
                }
            });

            return builder.build();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
    @NotNull
    public static String createConnectionToken() throws ConnectionTokenException {
        try {
            final Response<MyResponse> result = mService.getConnectionToken(MyApp.secretKey).execute();
            if (result.isSuccessful() && result.body() != null) {
                return (String)result.body().getData();
            } else {
                throw new ConnectionTokenException("Creating connection token failed");
            }
        } catch (IOException e) {
            throw new ConnectionTokenException("Creating connection token failed", e);
        }
    }

    public static Map<String,String> createPaymentIntent(String qrCode) throws IOException{
        Response<MyResponse> result = mService.createPaymentIntent(MyApp.secretKey,qrCode).execute();
        Log.d(MyApp.TERMINAL_TAG,new Gson().toJson(result.body()));
        if(result.body() != null && result.body().getCode() == 200){
            Object data = result.body().getData();
            if(data instanceof Map){
                Map<String,String> map = (Map<String,String>) data;
                return map;
            }
        }
        return null;
    }

    public static String lendPowerStripeTerminal(@NotNull Map<String,String> body) throws IOException {
        Response<MyResponse> result = mService.lendPowerStripeTerminal(body).execute();
        if(result.body() != null && result.body().getCode() == 200){
            return (String)result.body().getData();
        }
        return null;
    }

    public static Map<String,String> getPreAmount() throws IOException {
        final Response<MyResponse> result = mService.getPreAmount(MyApp.secretKey).execute();
        if (result.isSuccessful() && result.body() != null) {
            Map<String,String> data = (Map)result.body().getData();
            return data;
        }
        return null;
    }

    public static String lendPowerNayax(@NotNull Map<String,Object> body) throws IOException {
        Response<MyResponse> result = mService.lendPowerNayax(body).execute();
        if(result.body() != null && result.body().getCode() == 200){
            return (String)result.body().getData();
        }
        return null;
    }

    public static Map<String,Object> getVersion(String qrCode) throws IOException {
        final Response<MyResponse> result = mService.getVersion(qrCode).execute();
        if (result.isSuccessful() && result.body() != null) {
            Map<String,Object> data = (Map)result.body().getData();
            return data;
        }
        return null;
    }

    public static Map<String,Object> getBrightnessConfig(String qrCode) throws IOException {
        final Response<MyResponse> result = mService.getBrightnessConfig(qrCode).execute();
        if (result.isSuccessful() && result.body() != null) {
            Map<String,Object> data = (Map)result.body().getData();
            return data;
        }
        return null;
    }

    public static String getQrCode(@NotNull String fno) throws IOException {
        Response<MyResponse> result = mService.getQrCode(fno).execute();
        if(result.body() != null && result.body().getCode() == 200){
            Map<String,Object> data = (Map)result.body().getData();
            return (String)data.get("qrCode");
        }
        return null;
    }

    public static MyResponse getAD(@NotNull("qrCode") String qrCode, @NotNull("timestamp") Long timestamp, @NotNull("ip") String ip, @NotNull("sign") String sign) throws IOException {
        Response<MyResponse> result = mService.getAD(qrCode,timestamp,ip,sign).execute();
        return result.body();

    }

    public static MyResponse getLocationId(@NotNull String qrCode) throws IOException {
        Response<MyResponse> result = mService.getLocationId(qrCode).execute();
        return result.body();
    }

    public static MyResponse getOrderInfoByPowerBankId(@NotNull String powerBankId) throws IOException {
        Response<MyResponse> result = mService.getOrderInfoByPowerBankId(powerBankId).execute();
        return result.body();
    }
}

File: app/src/main/java/com/stwpower/powertap/data/api/MyApiClient.kt
package com.stwpower.powertap.data.api

import android.util.Log
import java.util.concurrent.TimeUnit
import com.google.gson.Gson
import com.stripe.stripeterminal.external.models.ConnectionTokenException
import com.stwpower.powertap.ConfigLoader
import com.stwpower.powertap.domain.MyResponse
import com.stwpower.powertap.utils.LoggingInterceptor
import java.io.IOException
import okhttp3.OkHttpClient
import retrofit2.*
import retrofit2.converter.gson.GsonConverterFactory
import java.security.SecureRandom
import javax.net.ssl.*
import java.security.cert.X509Certificate

object MyApiClient {

    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(20, TimeUnit.SECONDS)
        .writeTimeout(20, TimeUnit.SECONDS)
        .addInterceptor(LoggingInterceptor())
        .build()

    private val retrofit: Retrofit = Retrofit.Builder()
        .baseUrl(ConfigLoader.apiUrl + "/")
        .client(getUnsafeOkHttpClient())
        .addConverterFactory(GsonConverterFactory.create())
        .build()

    private val service: MyApiService = retrofit.create(MyApiService::class.java)

    fun getUnsafeOkHttpClient(): OkHttpClient {
        return try {
            val trustAllCerts = arrayOf<TrustManager>(
                object : X509TrustManager {
                    override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {}
                    override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {}
                    override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
                }
            )

            val sslContext = SSLContext.getInstance("SSL").apply {
                init(null, trustAllCerts, SecureRandom())
            }
            val sslSocketFactory = sslContext.socketFactory

            OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(20, TimeUnit.SECONDS)
                .writeTimeout(20, TimeUnit.SECONDS)
                .addInterceptor(LoggingInterceptor())
                .sslSocketFactory(sslSocketFactory, trustAllCerts[0] as X509TrustManager)
                .hostnameVerifier { _, _ -> true }
                .build()

        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    @Throws(ConnectionTokenException::class)
    fun createConnectionToken(): String {
        return try {
            val result = service.getConnectionToken(ConfigLoader.secretKey).execute()
            if (result.isSuccessful && result.body() != null) {
                result.body()!!.data as String
            } else {
                throw ConnectionTokenException("Creating connection token failed")
            }
        } catch (e: IOException) {
            throw ConnectionTokenException("Creating connection token failed", e)
        }
    }

    @Throws(IOException::class)
    fun createPaymentIntent(qrCode: String): Map<String, String>? {
        val result = service.createPaymentIntent(ConfigLoader.secretKey, qrCode).execute()
        Log.d("terminal", Gson().toJson(result.body()))
        if (result.body()?.code == 200) {
            val data = result.body()?.data
            if (data is Map<*, *>) {
                @Suppress("UNCHECKED_CAST")
                return data as Map<String, String>
            }
        }
        return null
    }

    @Throws(IOException::class)
    fun lendPowerStripeTerminal(body: Map<String, String>): String? {
        val result = service.lendPowerStripeTerminal(body).execute()
        return if (result.body()?.code == 200) {
            result.body()?.data as? String
        } else null
    }

    @Throws(IOException::class)
    fun getPreAmount(): Map<String, String>? {
        val result = service.getPreAmount(ConfigLoader.secretKey).execute()
        return if (result.isSuccessful && result.body() != null) {
            @Suppress("UNCHECKED_CAST")
            result.body()?.data as? Map<String, String>
        } else null
    }

    @Throws(IOException::class)
    fun lendPowerNayax(body: Map<String, Any>): String? {
        val result = service.lendPowerNayax(body).execute()
        return if (result.body()?.code == 200) {
            result.body()?.data as? String
        } else null
    }

    @Throws(IOException::class)
    fun getVersion(qrCode: String): Map<String, Any>? {
        val result = service.getVersion(qrCode).execute()
        return if (result.isSuccessful && result.body() != null) {
            @Suppress("UNCHECKED_CAST")
            result.body()?.data as? Map<String, Any>
        } else null
    }

    @Throws(IOException::class)
    fun getBrightnessConfig(qrCode: String): Map<String, Any>? {
        val result = service.getBrightnessConfig(qrCode).execute()
        return if (result.isSuccessful && result.body() != null) {
            @Suppress("UNCHECKED_CAST")
            result.body()?.data as? Map<String, Any>
        } else null
    }

    @Throws(IOException::class)
    fun getQrCode(fno: String): String? {
        android.util.Log.d("MyApiClient", "=== 开始调用getQrCode API ===")
        android.util.Log.d("MyApiClient", "请求参数 fno: $fno")

        try {
            val result = service.getQrCode(fno).execute()
            android.util.Log.d("MyApiClient", "API调用完成，HTTP状态码: ${result.code()}")
            android.util.Log.d("MyApiClient", "响应体: ${result.body()}")

            if (result.body()?.code == 200) {
                @Suppress("UNCHECKED_CAST")
                val data = result.body()?.data as? Map<String, Any>
                val qrCode = data?.get("qrCode") as? String
                android.util.Log.d("MyApiClient", "成功获取QR码: $qrCode")
                return qrCode
            } else {
                android.util.Log.w("MyApiClient", "API返回错误，code: ${result.body()?.code}, message: ${result.body()?.message}")
            }
        } catch (e: Exception) {
            android.util.Log.e("MyApiClient", "调用getQrCode API时发生异常", e)
            throw e
        }

        android.util.Log.w("MyApiClient", "getQrCode返回null")
        return null
    }

    @Throws(IOException::class)
    fun getAD(qrCode: String, timestamp: Long, ip: String, sign: String): MyResponse? {
        val result = service.getAD(qrCode, timestamp, ip, sign).execute()
        return result.body()
    }

    @Throws(IOException::class)
    fun getLocationId(qrCode: String): MyResponse? {
        val result = service.getLocationId(qrCode).execute()
        return result.body()
    }

    @Throws(IOException::class)
    fun getOrderInfoByPowerBankId(powerBankId: String): MyResponse? {
        val result = service.getOrderInfoByPowerBankId(powerBankId).execute()
        return result.body()
    }
}
